### PR TITLE
Implement MAV_CMD_REQUEST_MESSAGE

### PR DESCRIPTION
**Pull request by dayjaby**
**Describe problem solved by this pull request**
Implementing the MAV_CMD_REQUEST_MESSAGE as described in https://github.com/PX4/Firmware/issues/11441. So far only STORAGE_INFORMATION is supported.

I would like to use the opportunity and remind about one other PR, where all mavlink streams get their own file instead of having one bloated mavlink_messages.cpp: https://github.com/PX4/Firmware/pull/14259/commits/c1175e0d54cf60a655d2ce4db0323838a596a9cd
This would be helpful to get merged, otherwise I have double work here.

**Describe your solution**
When there is a stream with the same message ID, send the message once. Otherwise create a new stream, send the message once and delete the stream again.

**Test data / coverage**
Tested with QGC custom command file https://pastebin.com/maHNSQ1A:
image

### CODE REVIEW

**Comment by hamishwillee** [comment]
> Implementing the [MAV_CMD_REQUEST_MESSAGE](https://mavlink.io/en/messages/common.html#MAV_CMD_REQUEST_MESSAGE) as described in #11441. So far only STORAGE_INFORMATION is supported.

Not strictly speaking true. The request message has a default implementation that just immediately sends the last message if a stream has been defined for the message. So provided a message has a stream already defined then the "default" will be sent out.

But yes, this does need work . Firstly because there are cases like storage information and CAMERA_IMAGE_CAPTURED that take parameters (so a specific implementation required). Secondly because the current implementation I think ignores param 7, which is supposed to set the destination of the message. If you could fix those problems then that would be awesome.

**Comment by dayjaby** [comment]
I meant "so far" as in the scope of this PR ;)

> so you may be able to undo the specific changes you made for AUTOPILOT_VERSION etc.

what do you mean? Defining streams for those messages does make sense. It allows not only to use MAV_CMD_REQUEST_MESSAGE to request them, but also MAV_CMD_SET_MESSAGE_INTERVAL. Documenting it otherwise would be hard: "You can only request this particular subset of messages, but not set the interval for them".

Still trying to figure out the purpose of param7. For reference: https://github.com/mavlink/mavlink/issues/1248

**Comment by hamishwillee** [comment]
Hi @dayjaby 

>> so you may be able to undo the specific changes you made for AUTOPILOT_VERSION etc.
> what do you mean? Defining streams for those messages does make sense. It allows not only to use MAV_CMD_REQUEST_MESSAGE to request them, but also MAV_CMD_SET_MESSAGE_INTERVAL. Documenting it otherwise would be hard: "You can only request this particular subset of messages, but not set the interval for them".

Apologies. I had assumed these were already streamed - so you were just adding request message method to support the single request (which you would not need to do in most cases given the default implementation).  Ignore me. 

Re param7, as you say some context in mavlink/mavlink#1248 and https://github.com/mavlink/mavlink/pull/1249. Essentially most messages are broadcast so they go to everyone. But some messages have target addresses - they are intended for specific recipients. So for example if I'm telling something I want logging messages, then I'm the only thing that cares about those and the logging message should have my address. Honestly, in most cases I doubt it will matter if these were also broadcast, but setting this parameter allows you to choose whether you respond to the requestor or everyone. HOpe that helps. We can chat verbally if you want via slack at some point - just ping me (I'm in AEST so right now is 5pm on Monday)

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -500,30 +429,34 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
 	} else if (cmd_mavlink.command == MAV_CMD_GET_MESSAGE_INTERVAL) {
 		get_message_interval((int)roundf(cmd_mavlink.param1));
 
-	} else if (cmd_mavlink.command == MAV_CMD_REQUEST_FLIGHT_INFORMATION) {
-		send_flight_information();
-
-	} else if (cmd_mavlink.command == MAV_CMD_REQUEST_STORAGE_INFORMATION) {
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp:506]: We will have to support these commands for backwards compatibility, I suppose.

Comment (hamishwillee) [reply src/modules/mavlink/mavlink_receiver.cpp:506]: Yes we will. This is why we need microservices versioning and the new command metadata stuff.

**Comment by julianoes** [comment]
@dayjaby I hope it's ok that I just force-pushed to this PR and added some commits to support legacy commands. I'll do another iteration to try to address @bkueng's comments.

**Comment by dayjaby** [comment]
Looking great! So next step would be to add the broadcasting according to param7, right?

**Comment by julianoes** [comment]
> Looking great! So next step would be to add the broadcasting according to param7, right?

I have not looked at that one. Let me check what this involves.

**Comment by julianoes** [comment]
Hm. The broadcast thing seems very hard to implement because were just triggering the stream to send a message. Also, all the messages that I was just looking at are broadcast anyway.

**Comment by julianoes** [comment]
@bkueng I would appreciate your review.

**Comment by dayjaby** [comment]
> Hm. The broadcast thing seems very hard to implement because were just triggering the stream to send a message. Also, all the messages that I was just looking at are broadcast anyway.

But if the stream is not configured for other mavlink instances (or mavlink channel? what's the correct wording?), this code won't send them via these instances.

**Comment by julianoes** [comment]
> But if the stream is not configured for other mavlink instances (or mavlink channel? what's the correct wording?), this code won't send them via these instances.

Correct. It won't send it. I would say it's a limitation that we have for now. We basically don't support it as required by the spec. I don't see a trivial way to enable that right now, and I don't have a use-case, so I'm reluctant to do so.

@hamishwillee any suggestions how to proceed?

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 2]
```diff
@@ -588,6 +512,39 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
...
 }
 
+uint8_t MavlinkReceiver::handle_request_message_command(uint16_t message_id, float param2, float param3, float param4,
+		float param5, float param6, float param7)
+{
+	bool stream_found = false;
+
+	for (const auto stream : _mavlink->get_streams()) {
+		if (stream->get_id() == message_id) {
+			stream_found = stream->request_message(param2, param3, param4, param5, param6, param7);
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/mavlink/mavlink_receiver.cpp]: Addressed in 403dcd88a6717f36be8be9301f23f9c1836736e7.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 3]
```diff
@@ -588,6 +512,39 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
 	}
 }
 
+uint8_t MavlinkReceiver::handle_request_message_command(uint16_t message_id, float param2, float param3, float param4,
+		float param5, float param6, float param7)
+{
+	bool stream_found = false;
+
+	for (const auto stream : _mavlink->get_streams()) {
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/mavlink/mavlink_receiver.cpp]: Fixed in 1ec5d5df4b8cc224d5ed2e4421a454acf2b70404.

File: src/modules/mavlink/streams/flight_information.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	explicit MavlinkStreamFlightInformation(Mavlink *mavlink) : MavlinkStream(mavlink)
+	{}
+
+	bool send(const hrt_abstime t) override
+	{
+		actuator_armed_s actuator_armed{};
+		bool ret = _armed_sub.copy(&actuator_armed);
+
+		if (ret && actuator_armed.timestamp != 0) {
+			const param_t param_com_flight_uuid = param_find("COM_FLIGHT_UUID");
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/mavlink/streams/flight_information.h]: Fixed in 1ec5d5df4b8cc224d5ed2e4421a454acf2b70404.

File: src/modules/mavlink/streams/storage_information.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	bool request_message(float param2, float param3, float param4,
+			     float param5, float param6, float param7) override
+	{
+		storage_id = (int)roundf(param2);
+		return send(hrt_absolute_time());
+	}
+private:
+	int storage_id = 0;
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/mavlink/streams/storage_information.h]: Fixed in 1ec5d5df4b8cc224d5ed2e4421a454acf2b70404.

File: src/modules/mavlink/streams/autopilot_version.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#ifndef MAVLINK_STREAM_AUTOPILOT_VERSION_H
+#define MAVLINK_STREAM_AUTOPILOT_VERSION_H
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/mavlink/streams/autopilot_version.h]: Fixed in cce410adaf.

**Comment by julianoes** [comment]
Thanks for the review @bkueng. It should all be resolved now.

**Comment by hamishwillee** [comment]
>>> But if the stream is not configured for other mavlink instances (or mavlink channel? what's the correct wording?), this code won't send them via these instances.

>> Correct. It won't send it. I would say it's a limitation that we have for now. We basically don't support it as required by the spec. I don't see a trivial way to enable that right now, and I don't have a use-case, so I'm reluctant to do so.

> @hamishwillee any suggestions how to proceed?

I THINK you're saying that if a message is required to be broadcast it should be sent out on all channels, but our implementation only streams out on the particular channel that received the message. So for example, if the `MAV_CMD_REQUEST_MESSAGE` came from a companion computer connected by serial the response would go to the companion, but not (say) to a GCS connected by Wifi.

Is that correct?

If so, IMO:
1. This is a more general routing "problem" than this particular API. We should document that PX4 does not automatically route between interfaces/channels. 
2. Yes, it is fine that we don't try fix this here.
3. What we must do though is respect the param 7 values being set to 0 (broadcast) or 1 (the target recipient addresses).
4. The places I see potential probems we would have been a problem anyway. In practise I think that this should work in most cases.

**Comment by julianoes** [comment]
> So for example, if the MAV_CMD_REQUEST_MESSAGE came from a companion computer connected by serial the response would go to the companion, but not (say) to a GCS connected by Wifi.

Correct, assuming the companion and wifi are on separate MAVLink instances/ports in PX4 and not split at a later stage (e.g. in mavlink-router).

1. Correct, it doesn't automatically but it can forward from one to another instance if "forwarding" is set ([MAV_0_FORWARD](https://dev.px4.io/master/en/advanced/parameter_reference.html#MAV_0_FORWARD))
3. That would also be a lot of work. It would have to be implemented for each and every message, I think.

Thinking more about this I yet would want to see an actual use case where someone requests a message that they need sent out to someone else. I can't think of a case right now.

**Comment by julianoes** [comment]
@bkueng are you ok if we merge this. Everything you raised should be fixed.

**Comment by hamishwillee** [comment]
> Thinking more about this I yet would want to see an actual use case where someone requests a message that they need sent out to someone else. I can't think of a case right now.

Param updates emit PARAM_VALUE whenever any GCS changes a parameter, allowing all of them (in theory) to track changes. If two GCS were communicating on different channels then other GCS would not get the update.

Somewhat contrived, but this is the reason that broadcast routing says "send everywhere" is because you don't know what needs that message or on what channel it might live. But (also as already noted), if this sort of thing was a real /significant problem, we'd have addressed it already.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 7]
```diff
@@ -588,6 +512,41 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
...
+			for (const auto &stream : _mavlink->get_streams()) {
+				if (stream->get_id() == message_id) {
+					message_sent = stream->request_message(param2, param3, param4, param5, param6, param7);
+					break;
+				}
+			}
+		}
+	}
+
+	return (message_sent ? vehicle_command_ack_s::VEHICLE_RESULT_ACCEPTED : vehicle_command_ack_s::VEHICLE_RESULT_DENIED);
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_receiver.cpp:546]: This is a bit unfortunate, as you get a denied when the topic was not updated. It's a rare case but we'll have to see if it causes issues.

Comment (julianoes) [reply src/modules/mavlink/mavlink_receiver.cpp:546]: Right, that's why we have to use `copy()` for some of the ones that we expect to use like this:
https://github.com/PX4/Firmware/pull/14815/files/b0ed9d5ca7f1d206fbdae2c0b415e90859bf2ba6#diff-c8466a5bb81a917a3cf7b4131c18f821R89

**Comment by julianoes** [comment]
@bkueng how would you cleanup these commits exactly? I'm not sure what's best. I would rebase them as is.

**Comment by bkueng** [comment]
I'd squash them if you don't find a good way to split (I don't see one either). 18 commits is just unnecessary for this.

**Comment by julianoes** [comment]
Selection_005

It just makes me a bit sad squashing this when it would actually help the next reader.

I'd say it makes sense to squash all fixup commits, and then rebase the commits.

**Comment by LorenzMeier** [comment]
TLDR; I think the number of commits you've chosen makes perfectly sense, let's just not merge fixup commits.

I generally would advise doing rather more commits than less, but yes, please do not keep fixup commits in the history. A good rule of thumb is 20-50 lines of code changed in one consistent set of files per commit. So a couple days work should be 15-50 commits.

**Comment by julianoes** [comment]
Ok. It's just that getting rid of the fixups is not trivial since it's adding headers to files that @dayjaby had added. So I could go back and add them back in where it belongs but it would take a bit of time for not much gain.

**Comment by julianoes** [comment]
Alright, I think it's too much effort to perfectly rewrite history here and I see value (and no big problems) with the current commits, so I'm going to merge it. Thanks for the reviews.
